tokio/src/util/linked_list.rs

```diff
@@ -294,6 +294,32 @@ cfg_io_readiness! {
     }
 }
 
+cfg_taskdump! {
+    impl<T: Link> LinkedList<T, T::Target> {
+        pub(crate) fn for_each<F>(&mut self, mut f: F)
```

Comment:
It isn't obvious why this method uses `pop_`. It looks like iteration, why is it mutating the list to iterate??

Reply:
I took a stab at writing an iterator over `LinkedList` and it segfaulted — so I shifted gears to see if I could implement `for_each` completely safely instead using the existing API.

Reply:
Please try this:
```rs
let mut next = self.head;

while let Some(curr) = next {
    unsafe {
        f(&mut *curr.as_ptr());
        next = T::pointers(curr).as_ref().get_next();
    }
}
```

Reply:
@Darksonn `&*curr.as_ptr()` gives a `&T::Target`, but `f` consumes a `&T::Handle`.

Reply:
Okay, try this:
```Rust
let mut next = self.head;

while let Some(curr) = next {
    unsafe {
        let handle = ManuallyDrop::new(T::from_raw(curr));
        f(&handle);
        next = T::pointers(curr).as_ref().get_next();
    }
}
```